impl/src/main/java/org/jboss/weld/bean/attributes/BeanAttributesFactory.java

```diff
@@ -171,7 +171,7 @@ protected void initQualifiers(Set<Annotation> qualifiers) {
             if (qualifiers.isEmpty()) {
                 this.qualifiers = DEFAULT_QUALIFIERS;
             } else {
-                ArraySet<Annotation> normalizedQualifiers = new ArraySet<Annotation>(qualifiers.size() + 2);
+                Set<Annotation> normalizedQualifiers = new HashSet<Annotation>(qualifiers.size() + 2);
```

Comment:
Yes, that's the point. Either use `ImmutableSet.copyOf()` or modify the algorithm not to use `remove()` ;-) 

`if (qualifiers.size() == 1) {` could be also "merged" with the nested condition.